src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs

```diff
@@ -307,128 +402,148 @@ public static (uint encoded, int bitLength) Encode(int data)
         /// <returns>The number of decoded symbols.</returns>
         public static int Decode(ReadOnlySpan<byte> src, ref byte[] dstArray)
         {
+            // The code below implements the decoding logic for a HPack huffman encoded literal values.
+            // https://httpwg.org/specs/rfc7541.html#string.literal.representation
+            //
+            // To decode a symbol, we traverse the decoding lookup table tree by 8 bits for each lookup
+            // until we found a leaf - which contains decoded symbol (octet)
+            //
+            // see comments in GenerateDecodingLookupTree() describing decoding table
+
             Span<byte> dst = dstArray;
             Debug.Assert(dst != null && dst.Length > 0);
 
+            if (s_decodingTree == null)
+            {
+                Volatile.Write(ref s_decodingTree, GenerateDecodingLookupTree());
+            }
+            ushort[] decodingTree = s_decodingTree;
+
+            int lookupTableIndex = 0;
+            int lookupIndex;
+
+            ulong acc = 0;
```

Comment:
Is there a reason for 64bit `acc`? Is it perf? Or am I missing something fundamental about the algorithm? :smile:

Reply:
It would work for 16 bits and 32 bits acc as well. By having longest possible accumulator for most common OS (64 bits) shifting operations to get next 8 bits is very well compiled into Amd64 machine code by JIT. Same would apply for 32 bits acc, but the bigger the acc the less transition between loading and decoding data would occur.
Those would be line 431 & 480.
However, if I would read it by bytes I could safe lines 434 & 439 !
I will do some perf test for 16 bit accumulator with eliminating loop at line 439
I'll be back :-)

Reply:
It turned out, after running benchmarks, that reading just next byte into UInt16 is actually faster, with ration of 0.93, and, to me, more readable. Additionally it eliminates suboptimal performance of dealing with ulong in 32 bit OS.
Have changed it.

|      Method | DataIndex |         Mean |     Error |    StdDev |         Op/s | Ratio |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------ |---------- |-------------:|----------:|----------:|-------------:|------:|-------:|------:|------:|----------:|
|     Current |        -1 |    575.93 ns |  4.116 ns |  3.648 ns |  1,736,315.4 |  1.00 |      - |     - |     - |      64 B |
| Acc16bits |        -1 |    536.60 ns |  3.289 ns |  3.076 ns |  1,863,588.2 |  0.93 |      - |     - |     - |      64 B |